### PR TITLE
Changes rationale for Object.assign

### PR DESCRIPTION
**Pull request by bennypowers**
object spread is not just syntax sugar but actually performs better in browsers. The "When Not To Use It" section should therefore give a more nuanced opinion on when Object.assign is desireable.

**What is the purpose of this pull request? (put an "X" next to item)**

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**
Changed the "When Not To Use It" section of the `prefer-object-spread` docs.

**Is there anything you'd like reviewers to focus on?**
More concrete ideas for positive uses for `Object.assign`

### CODE REVIEW

**Comment by platinumazure** [comment]
Hi @bennypowers, thanks for the PR.

The new wording feels a little opinionated to me. Wondering if it might be better just to suggest that the rule could be disabled when object spread is not available for the user's codebase?

**Comment by bennypowers** [comment]
@platinumazure I'm ok with that in principle: my main issue was the phrase "syntactic sugar" which implied that both forms are entirely equivalent.

OTOH I've often learned nice little tidbits from eslint docs, which is why I wanted to offer users a more concrete example of the advantages of the assign method

**Comment by platinumazure** [comment]
@bennypowers Fair enough. I'm wondering if maybe the introduction of the rule could be a good place to go into the benefits of object spread vs `Object.assign`? What do you think?

**Comment by bennypowers** [comment]
@platinumazure it's my first rodeo in eslint docs land, so I'll defer to your judgement.

Will revise for your review a little later on.

**Comment by bennypowers** [comment]
Ok I updated the wording below and added some context above. Please review and let me know if this is more in line with what you had in mind.

File: docs/rules/prefer-object-spread.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 When Object.assign is called using an object literal as the first argument, this rule requires using the object spread syntax instead. This rule also warns on cases where an `Object.assign` call is made using a single argument that is an object literal, in this case, the `Object.assign` call is not needed.
 
+Object spread is a declarative alternative which may perform better than the more dynamic, imperative `Object.assign`
```
Comment (platinumazure) [inline docs/rules/prefer-object-spread.md]: Could you please add a period to the end of this sentence?

Comment (bennypowers) [reply docs/rules/prefer-object-spread.md]: good catch. pushed.

**Comment by ljharb** [comment]
The wording is fine, but performance is entirely unrelated to whether it’s syntax sugar.

**Comment by platinumazure** [comment]
@bennypowers I apologize for the delay in reviewing again. My work week was terrible and I didn't have much time for open-source activities.

@ljharb Could you please clarify your comment?

**Comment by ljharb** [comment]
> object spread is not just syntax sugar but actually performs better in browsers

Something being syntax sugar is unrelated to its performance. I would rephrase this as “object spread should be preferred not just because it’s clearer syntax sugar but also because it performs better in browsers”

**Comment by bennypowers** [comment]
I'm cool with @ljharb phrasing. Can we remove the word "sugar"? In a speculative 5-years-from-now, I could imagine spread considered the idiom, and Object.assign a specialty function.

**Comment by ljharb** [comment]
It's still forever going to be syntax sugar - since it's syntax that replicates the functionality of API. It's not called sugar because of being idiomatic, it's an actual definition.

**Comment by platinumazure** [comment]
@ljharb You're quoting from the initial issue post. Do you have any issue with the actual commit?

**Comment by ljharb** [comment]
@platinumazure nope, but I think it's important to clarify the well-understood definition of "syntax sugar" in the PR, for future readers :-) the commit [is fine](
https://github.com/eslint/eslint/pull/10621#issuecomment-406812026).

**Comment by platinumazure** [comment]
Thanks for clarifying @ljharb.

With that, I'm going to go ahead and merge this. Thanks @bennypowers for contributing!
